### PR TITLE
Port late days and homework extensions

### PR DESCRIPTION
**Pull request by MarisaPLee**

### CODE REVIEW

File: site/app/libraries/database/DatabaseQueriesPostgresql.php [hunk 1]
```diff
@@ -892,4 +898,71 @@ public function getTeamByUserId($g_id, $user_id) {
...
+            AND g_id=?;", array($days, $user_id, $g_id));
+        if(count($this->database->rows())==0){
+          $this->database->query("
+            INSERT INTO late_day_exceptions
+            (user_id, g_id, late_day_exceptions)
+            VALUES(?,?,?)", array($user_id, $g_id, $days));
+        }
+    }
 }
+
```
Comment (MasterOdin) [inline site/app/libraries/database/DatabaseQueriesPostgresql.php]: Add function definitions to app/libraries/IDatabaseQueries.php please

File: site/app/controllers/admin/LateController.php [hunk 2]
```diff
@@ -0,0 +1,232 @@
...
+    }
+
+    function validate_homework($id) {
+        $g_ids = $this->core->getQueries()->getAllElectronicGradeablesIds();
+            foreach($g_ids as $index => $value) {
+                if($id == $value[0]){
+                    return true;
+                }
+            }
+        return false;
```
Comment (MasterOdin) [inline site/app/controllers/admin/LateController.php]: fix indentation, make function private, use lowerCamelCase

File: site/app/controllers/admin/LateController.php [hunk 3]
```diff
@@ -0,0 +1,232 @@
...
+            $data = null;
+            return false;
+        }
+        foreach($rows as $row) {
+
+            $fields = explode(',', $row);
+            //Remove any extraneous whitespace at beginning/end of all fields.
+            foreach($fields as &$field) {
+                $field = trim($field);
+            } unset($field);
```
Comment (MasterOdin) [inline site/app/controllers/admin/LateController.php]: replace with `$fields = array_map(function($k) { return trim($k); }, $fields);`

File: site/app/controllers/admin/LateController.php [hunk 4]
```diff
@@ -0,0 +1,232 @@
...
+        foreach($users as $user){
+            $user_table[] = array('user_id' => $user->getId(),'user_firstname' => $user->getDisplayedFirstName(), 'user_lastname' => $user->getLastName(), 'late_day_exceptions' => $user->getLateDayExceptions());
+        }
+        $this->core->getOutput()->renderJson(array(
+            'gradeable_id' => $g_id,
+            'users' => $user_table
+        ));
+    }
+
+    function parse_and_validate_csv($csv_file, &$data, $type) {
```
Comment (MasterOdin) [inline site/app/controllers/admin/LateController.php]: mark as private and use lowerCamelCase for name

File: site/app/controllers/admin/LateController.php [hunk 5]
```diff
@@ -0,0 +1,232 @@
...
+        case ($tmp[0] && $tmp[0]->format('m-d-Y') === $timestamp):
+        case ($tmp[1] && $tmp[1]->format('m/d/Y') === $timestamp):
+        case ($tmp[2] && $tmp[2]->format('m-d-y') === $timestamp):
+        case ($tmp[3] && $tmp[3]->format('m/d/y') === $timestamp):
+            return true;
+        default:
+            return false;
+        }
+        return true;
+    }
```
Comment (MasterOdin) [inline site/app/controllers/admin/LateController.php]: move this function to /app/libraries/DateUtils.php and mark ti static and public, use lowerCamelCase

File: site/app/models/SimpleLateUser.php [hunk 6]
```diff
@@ -0,0 +1,58 @@
...
+class SimpleLateUser extends AbstractModel {
+    
+    /** @property @var bool Is this user actually loaded (else you cannot access the other member variables) */
+    protected $loaded = false;
+    
+    /** @property @var string The id of this user which should be a unique identifier (ex: RCS ID at RPI) */
+    protected $id;
+    /** @property @var string The first name of the user */
+    protected $first_name;
+    /** @property @var string The first name of the user */
```
Comment (wangj35) [inline site/app/models/SimpleLateUser.php]: for the comment, insert preferred before first name

**Comment by bmcutler** [comment]
@MarisaPLee can you resolve the merge conflicts?  Then we'll merge.

**Pull request merged**